include/deal.II/matrix_free/tools.h

```diff
@@ -509,6 +567,153 @@ namespace MatrixFreeTools
       false);
   }
 
+  template <int dim,
+            int fe_degree,
+            int n_q_points_1d,
+            int n_components,
+            typename Number,
+            typename VectorizedArrayType,
+            typename MatrixType>
+  void
+  compute_matrix(
+    const MatrixFree<dim, Number, VectorizedArrayType> &            matrix_free,
+    const AffineConstraints<Number> &                               constraints,
+    MatrixType &                                                    matrix,
+    const std::function<void(FEEvaluation<dim,
+                                          fe_degree,
+                                          n_q_points_1d,
+                                          n_components,
+                                          Number,
+                                          VectorizedArrayType> &)> &local_vmult,
+    const unsigned int                                              dof_no,
+    const unsigned int                                              quad_no,
+    const unsigned int first_selected_component)
+  {
+    matrix_free.template cell_loop<MatrixType, MatrixType>(
+
+      [&](const auto &, auto &dst, const auto &, const auto range) {
+        FEEvaluation<dim,
+                     fe_degree,
+                     n_q_points_1d,
+                     n_components,
+                     Number,
+                     VectorizedArrayType>
+          integrator(matrix_free, range);
+
+        unsigned int const dofs_per_cell = integrator.dofs_per_cell;
+
+        for (auto cell = range.first; cell < range.second; ++cell)
+          {
+            unsigned int const n_filled_lanes =
+              matrix_free.n_active_entries_per_cell_batch(cell);
+
+            FullMatrix<TrilinosScalar>
+              matrices[VectorizedArray<double>::size()];
+
+            std::fill_n(matrices,
+                        VectorizedArray<double>::size(),
+                        FullMatrix<TrilinosScalar>(dofs_per_cell,
+                                                   dofs_per_cell));
+
+            integrator.reinit(cell);
+
+            for (unsigned int j = 0; j < dofs_per_cell; ++j)
+              {
+                for (unsigned int i = 0; i < integrator.dofs_per_cell; ++i)
+                  integrator.begin_dof_values()[i] =
+                    static_cast<double>(i == j);
+
+                local_vmult(integrator);
+
+                for (unsigned int i = 0; i < dofs_per_cell; ++i)
+                  for (unsigned int v = 0; v < n_filled_lanes; ++v)
+                    matrices[v](i, j) = integrator.begin_dof_values()[i][v];
+              }
+
+
+            // finally assemble local matrices into global matrix
+            for (unsigned int v = 0; v < n_filled_lanes; v++)
+              {
+                auto cell_v = matrix_free.get_cell_iterator(cell, v);
+
+                std::vector<types::global_dof_index> dof_indices(dofs_per_cell);
```

Comment:
I am not sure whether it matters at this level, but I tend to prefer moving the allocation of `dof_indices` (and `temp` below) outside of this loop (i.e., line 604) so the memory allocation gets amortized a bit. (Now, to be honest, at least for the case of complicated constraints there is a memory allocation happening inside `constraints.distribute_local_to_global`, so it won't matter in the grand scheme.)